### ISSUE TITLE
Announce events not fired

### ISSUE THREAD

**Issue by klimenttoshkov** [zerodahero/laravel-workflow#34]
Hi,

I have the following test model:
```
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use ZeroDaHero\LaravelWorkflow\Traits\WorkflowTrait;

class Test extends Model
{
    use HasFactory;
    use WorkflowTrait;

    protected $table = 'test';

    protected $casts = [
        'marking' => 'array',
    ];


}
```

Workflow:
```
'test' => [
        'type' => 'workflow',
        // 'marking_store' => [
        //     'type' => 'multiple_state',
        // ],
        'supports' => [
            Test::class,
        ],

        'events_to_dispatch' => [
            'workflow.announce',
        ],

        'places' => [
            'a', 'b', 'c',
        ],

        'initial_places' => [
            'a',
        ],

        'transitions' => [
            't1' => [
                'from' => 'a',
                'to' => 'b'
            ],
            't2' => [
                'from' => 'b',
                'to' => 'c'
            ],

        ],
    ];
```

Event subscriber:
```
<?php

namespace App\Listeners;

use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Support\Facades\Log;
use Symfony\Component\Workflow\Event\AnnounceEvent;
use ZeroDaHero\LaravelWorkflow\Events\GuardEvent;

class TestSubscriber implements ShouldQueue
{

    public function handleAnnounce($type, $event)
    {
        Log::debug('* Received Event: ' . $type);

        Log::debug(gettype($event));
    }


    public function subscribe($events)
    {
        $events->listen(
            'workflow.test.*',
            [TestSubscriber::class, 'handleAnnounce']
        );
    }
}
```

And the following testing code:
```
        Log::debug('---');
        $test = new Test();

        Log::debug('Checking transition t1: ' . $test->workflow_can('t1'));
        Log::debug('Current marking: ' . $test->marking);

        foreach ($test->workflow_transitions() as $transition) {
            Log::debug('Enabled Transition: ' . $transition->getName());
        }

        Log::debug('Applying transition t1:');
        $marking = $test->workflow_apply('t1');

        return 0;
```

What I expect is to receive the event somehow so I can process it.

All of the above was discovered when I am trying to receive Announce events, in order to make automatic transitions when possible. So far I have discovered that they are not fired at all. This is issue 1.

According to current documentation, the handler in subscriber should receive Event as first argument. 
Actually it receives Event only in the case when subscriber is subscribed to ZeroDaHero Events.
If subscribed to Symfony dot-style events (as in the example above), function receives string containing the event name and second parameter of type array, that contains dump of the ZeroDaHero Event. This is issue 2.

I believe this is related to latest changes in Symfony/Workflow 5.2.

here is the debug information from above code:

```
[2021-01-21 16:22:57] local.DEBUG: ---  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.entered  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard.t1  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: Checking transition t1: 1  
[2021-01-21 16:22:57] local.DEBUG: Current marking: a  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard.t1  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: Enabled Transition: t1  
[2021-01-21 16:22:57] local.DEBUG: Applying transition t1:  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard.t1  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.leave  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.leave.a  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.transition  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.transition.t1  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.enter  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.enter.b  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.entered  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.entered.b  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.completed  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.completed.t1  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard  
[2021-01-21 16:22:57] local.DEBUG: array  
[2021-01-21 16:22:57] local.DEBUG: * Received Event: workflow.test.guard.t2  
[2021-01-21 16:22:57] local.DEBUG: array  
```

        "symfony/workflow": "^5.2",
        "zerodahero/laravel-workflow": "^3.2"

**Comment by zerodahero** [comment]
Hi @klimenttoshkov, thanks for the detailed info! That helped me pinpoint the issue very easily.

Looks like I missed the 'announce' event when it was added into the Symfony workflow component. I'll try and find some time this week to add that in.

As for issue 2, there's not much we can do about that since that's Laravel's way of handling events. It's the same if you listen for specific model or framework events: the "event" is the first argument (which is either the event name, or the event class/object) and the payload is second (which is the data present for the non-class events).

I'll let you know when I have a new version ready with the announce events dispatching.

**Comment by klimenttoshkov** [comment]
Already created pull request, maybe you can merge it ... 

In regard to events, I've been digging around and already found out the same thing. I will be exploring further though.

**Comment by klimenttoshkov** [comment]
PS. Could you please look into how to add support for         'events_to_dispatch' => [

**Comment by klimenttoshkov** [comment]
In regard to wildcard event listeners, I see that event data is passed as second argument in array. That is fine, but the array itself is event in kind of serialized form. Is this the intended behaviour?

How can I use this? Here is an example dump of second argument value:
```
Array
(
    [0] => ZeroDaHero\LaravelWorkflow\Events\AnnounceEvent Object
        (
            [originalEvent:protected] => Symfony\Component\Workflow\Event\AnnounceEvent Object
                (
                    [context:protected] => Array
                        (
                        )

                    [subject:Symfony\Component\Workflow\Event\Event:private] => App\Models\Test Object
                        (
                            [table:protected] => test
                            [casts:protected] => Array
                                (
                                    [marking] => array
                                )

                            [connection:protected] => 
                            [primaryKey:protected] => id
                            [keyType:protected] => int
                            [incrementing] => 1
                            [with:protected] => Array
                                (
                                )

                            [withCount:protected] => Array
                                (
                                )

                            [perPage:protected] => 15
                            [exists] => 
                            [wasRecentlyCreated] => 
                            [attributes:protected] => Array
                                (
                                    [marking] => "b"
                                )

                            [original:protected] => Array
                                (
                                )

                            [changes:protected] => Array
                                (
                                )

                            [classCastCache:protected] => Array
                                (
                                )

                            [dates:protected] => Array
                                (
                                )

                            [dateFormat:protected] => 
                            [appends:protected] => Array
                                (
                                )

                            [dispatchesEvents:protected] => Array
                                (
                                )

                            [observables:protected] => Array
                                (
                                )

                            [relations:protected] => Array
                                (
                                )

                            [touches:protected] => Array
                                (
                                )

                            [timestamps] => 1
                            [hidden:protected] => Array
                                (
                                )

                            [visible:protected] => Array
                                (
                                )

                            [fillable:protected] => Array
                                (
                                )

                            [guarded:protected] => Array
                                (
                                    [0] => *
                                )

                        )

                    [marking:Symfony\Component\Workflow\Event\Event:private] => Symfony\Component\Workflow\Marking Object
                        (
                            [places:Symfony\Component\Workflow\Marking:private] => Array
                                (
                                    [b] => 1
                                )

                        )

                    [transition:Symfony\Component\Workflow\Event\Event:private] => Symfony\Component\Workflow\Transition Object
                        (
                            [name:Symfony\Component\Workflow\Transition:private] => t1
                            [froms:Symfony\Component\Workflow\Transition:private] => Array
                                (
                                    [0] => a
                                )

                            [tos:Symfony\Component\Workflow\Transition:private] => Array
                                (
                                    [0] => b
                                )

                        )

                    [workflow:Symfony\Component\Workflow\Event\Event:private] => Symfony\Component\Workflow\Workflow Object
                        (
                            [definition:Symfony\Component\Workflow\Workflow:private] => Symfony\Component\Workflow\Definition Object
                                (
                                    [places:Symfony\Component\Workflow\Definition:private] => Array
                                        (
                                            [a] => a
                                            [b] => b
                                            [c] => c
                                        )

                                    [transitions:Symfony\Component\Workflow\Definition:private] => Array
                                        (
                                            [0] => Symfony\Component\Workflow\Transition Object
                                                (
                                                    [name:Symfony\Component\Workflow\Transition:private] => t1
                                                    [froms:Symfony\Component\Workflow\Transition:private] => Array
                                                        (
                                                            [0] => a
                                                        )

                                                    [tos:Symfony\Component\Workflow\Transition:private] => Array
                                                        (
                                                            [0] => b
                                                        )

                                                )

                                            [1] => Symfony\Component\Workflow\Transition Object
                                                (
                                                    [name:Symfony\Component\Workflow\Transition:private] => t2
                                                    [froms:Symfony\Component\Workflow\Transition:private] => Array
                                                        (
                                                            [0] => b
                                                        )

                                                    [tos:Symfony\Component\Workflow\Transition:private] => Array
                                                        (
                                                            [0] => c
                                                        )

                                                )

                                        )

                                    [initialPlaces:Symfony\Component\Workflow\Definition:private] => Array
                                        (
                                            [0] => a
                                        )

                                    [metadataStore:Symfony\Component\Workflow\Definition:private] => Symfony\Component\Workflow\Metadata\InMemoryMetadataStore Object
                                        (
                                            [workflowMetadata:Symfony\Component\Workflow\Metadata\InMemoryMetadataStore:private] => Array
                                                (
                                                )

                                            [placesMetadata:Symfony\Component\Workflow\Metadata\InMemoryMetadataStore:private] => Array
                                                (
                                                )

                                            [transitionsMetadata:Symfony\Component\Workflow\Metadata\InMemoryMetadataStore:private] => SplObjectStorage Object
                                                (
                                                    [storage:SplObjectStorage:private] => Array
                                                        (
                                                        )

                                                )

                                        )

                                )

                            [markingStore:Symfony\Component\Workflow\Workflow:private] => ZeroDaHero\LaravelWorkflow\MarkingStores\EloquentMarkingStore Object
                                (
                                    [singleState:ZeroDaHero\LaravelWorkflow\MarkingStores\EloquentMarkingStore:private] => 1
                                    [property:ZeroDaHero\LaravelWorkflow\MarkingStores\EloquentMarkingStore:private] => marking
                                )

                            [dispatcher:Symfony\Component\Workflow\Workflow:private] => Symfony\Component\EventDispatcher\EventDispatcher Object
                                (
                                    [listeners:Symfony\Component\EventDispatcher\EventDispatcher:private] => Array
                                        (
                                            [workflow.guard] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [0] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                        (
                                                                        )

                                                                    [1] => guardEvent
                                                                )

                                                        )

                                                )

                                            [workflow.leave] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [0] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                        (
                                                                        )

                                                                    [1] => leaveEvent
                                                                )

                                                        )

                                                )

                                            [workflow.transition] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [0] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                        (
                                                                        )

                                                                    [1] => transitionEvent
                                                                )

                                                        )

                                                )

                                            [workflow.enter] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [0] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                        (
                                                                        )

                                                                    [1] => enterEvent
                                                                )

                                                        )

                                                )

                                            [workflow.entered] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [0] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                        (
                                                                        )

                                                                    [1] => enteredEvent
                                                                )

                                                        )

                                                )

                                            [workflow.completed] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [0] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                        (
                                                                        )

                                                                    [1] => completedEvent
                                                                )

                                                        )

                                                )

                                            [workflow.announce] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [0] => Array
                                                                (
                                                                    [0] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                        (
                                                                        )

                                                                    [1] => announceEvent
                                                                )

                                                        )

                                                )

                                        )

                                    [sorted:Symfony\Component\EventDispatcher\EventDispatcher:private] => Array
                                        (
                                        )

                                    [optimized:Symfony\Component\EventDispatcher\EventDispatcher:private] => Array
                                        (
                                            [workflow.entered] => Array
                                                (
                                                    [0] => Closure Object
                                                        (
                                                            [this] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                (
                                                                )

                                                            [parameter] => Array
                                                                (
                                                                    [$event] => <required>
                                                                )

                                                        )

                                                )

                                            [workflow.guard] => Array
                                                (
                                                    [0] => Closure Object
                                                        (
                                                            [this] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                (
                                                                )

                                                            [parameter] => Array
                                                                (
                                                                    [$event] => <required>
                                                                )

                                                        )

                                                )

                                            [workflow.leave] => Array
                                                (
                                                    [0] => Closure Object
                                                        (
                                                            [this] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                (
                                                                )

                                                            [parameter] => Array
                                                                (
                                                                    [$event] => <required>
                                                                )

                                                        )

                                                )

                                            [workflow.transition] => Array
                                                (
                                                    [0] => Closure Object
                                                        (
                                                            [this] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                (
                                                                )

                                                            [parameter] => Array
                                                                (
                                                                    [$event] => <required>
                                                                )

                                                        )

                                                )

                                            [workflow.enter] => Array
                                                (
                                                    [0] => Closure Object
                                                        (
                                                            [this] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                (
                                                                )

                                                            [parameter] => Array
                                                                (
                                                                    [$event] => <required>
                                                                )

                                                        )

                                                )

                                            [workflow.completed] => Array
                                                (
                                                    [0] => Closure Object
                                                        (
                                                            [this] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                (
                                                                )

                                                            [parameter] => Array
                                                                (
                                                                    [$event] => <required>
                                                                )

                                                        )

                                                )

                                            [workflow.announce] => Array
                                                (
                                                    [0] => Closure Object
                                                        (
                                                            [this] => ZeroDaHero\LaravelWorkflow\Events\WorkflowSubscriber Object
                                                                (
                                                                )

                                                            [parameter] => Array
                                                                (
                                                                    [$event] => <required>
                                                                )

                                                        )

                                                )

                                        )

                                )

                            [name:Symfony\Component\Workflow\Workflow:private] => test
                            [eventsToDispatch:Symfony\Component\Workflow\Workflow:private] => 
                        )

                    [propagationStopped:Symfony\Contracts\EventDispatcher\Event:private] => 
                )

        )

)
```

**Comment by klimenttoshkov** [comment]
One more thing :-)

Announce event that is fired is for the already completed transition, not for the next available transition.

(A) -> [t1] -> (B) -> [t2] -> (C)

Apply(t1) is expected to receive at some point announce.t2 but instead announce.t1 is fired:
```
[2021-01-21 18:08:13] local.DEBUG: Applying transition t1:  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.guard  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.guard.t1  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.leave  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.leave.a  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.transition  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.transition.t1  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.enter  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.enter.b  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.entered  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.entered.b  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.completed  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.completed.t1  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.announce  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.announce.t1  
[2021-01-21 18:08:13] local.DEBUG: * Received Event: string  
[2021-01-21 18:08:13] local.DEBUG:  * Event: workflow.test.announce.t1  
[2021-01-21 18:08:13] local.DEBUG:    * b  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.guard  
[2021-01-21 18:08:13] local.DEBUG: -- Event: workflow.test.guard.t2  
```

Do I misunderstand something on conceptual level in announcements?

**Comment by zerodahero** [comment]
@klimenttoshkov I just pushed a few changes to `develop` that should take care of both the announce event and `events_to_dispatch` option. Whenever you get a chance, can you try that out and see if it resolves everything? I can push a release if everything is ok.

With the announce event, I noticed the same thing. It looks like from the Symfony source, the announce event "announces" the transition that _just completed_. The main difference between announce and completed though is that announce dispatches the `guard` event(s) for the _new marking_. So, the events you listened to above are correct, it's just that `workflow.test.announce.t1` actually fires the event for `workflow.test.guard.t2`. (So, with the `events_to_dispatch` option, if you remove "announce" you shouldn't see `workflow.test.guard.t2` anymore)

**Comment by klimenttoshkov** [comment]
OK, I've tested the development branch, here are the results.

1. Omitting workflow type in config results in "workflow" type as expected - OK
2. Omitting marking_store in config results in "multiple_state" as expected - OK
3. Specifying type "state_machine" does not render the workflow without squares for transitions (see image below)
4. Apparently specifying 'events_to_dispatch' => [] does not prevent all events from firing all of them. Also specifying some events does not limit the dispatched events to specified branch only. **All events are always dispatched** (including announce events as already fixed in previous patch)
5. Something that I did not bring to your attention, but I see is fixed now (maybe as consequence): when creating a model, it now enters the default place and appropriate events for it are dispatched. Previously the model was in the place but checking the marking returned empty string and only after checking available transitions enters the place.

![test](https://user-images.githubusercontent.com/59835221/105744529-f4d64f80-5f45-11eb-9e6c-2cb1fa03b1f0.png)

**Comment by klimenttoshkov** [comment]
Btw. I completely miss the logic of firing guard events for the next available transitions. What useful purpose that could serve?
I imagined I can make workflow to auto advance without explicitly applying transitions, but it does not seem the use case.

**Comment by zerodahero** [comment]
Ok, let me try and get a live test going on those events. I was able to silence them all in tests, so I'll have to look a little closer and see what I missed. Out of curiosity, how are you setting that value? Just on the config in the workflow?

Guard events will always fire still, the Symfony component itself doesn't consider whether those should be fired or not--as long as there's a dispatcher they fire.

I think if your workflow is linear (or can be constrained to be linear in certain sections), the guard events could be used to auto-advance the workflow. If they pass, they could apply the next transition. With the announce event, that would mean a subject could transition all the way through if unblocked by the guards.

**Comment by klimenttoshkov** [comment]
config/workflow.php
```
    'test' => [
        'type' => 'state_machine',
        // 'marking_store' => [
            // 'type' => 'single_state',
            // 'type' => 'multiple_state',
        // ],

        'metadata' => [
            'title' => 'Blog Publishing Workflow',
        ],

        'supports' => [
            Test::class,
        ],

        'events_to_dispatch' => [
            'workflow.announce',
        //     'completed',
            // 'announce',
        ],
```

I know about Guard events always firing up, it is mentioned in the documentation and this behaviour makes perfect sense for me.

In regard to using Guard to auto-advance I am confident this should be avoided at all cost. First business logic should not be put in Guard and second (most-important) is that all **Guards may be fired many times during the life cycle**. Having something there that makes changes to the Model will have unpredictable results. I know this last part of the conversation is general discussion, it is not directly related to your work. But I was asking do I miss something conceptual here. In my opinion the system should fire announce.t2 when t1 is completed.

Please share your comments on 5) above, I am curious how it was fixed. Before your last patches the model was in the default places but enters them silently without any events unless you check for possible transitions or current place.

Also please comment on that dumped image appears as Workflow instead of State Machine. At least according to Symfony documentation when dumping State Machine there should be labels for each transition above the arrow, instead of plotting them as squares.

Looking forward to seeing new release so I can update production. Thank you!

**Comment by klimenttoshkov** [comment]
Hi again,

Regarding to 4 - please disregard it completely.
I was testing with type state_machine. Upon reviewing your latest commit, I found out that state_machine is created without $dispatches events parameter, which cleared the situation for me. Even more, I repeated my tests with type = workflow and I confirm expected events are properly dispatched. Also noted the use of predefined constants in Symfony\Component\Workflow\WorkflowEvents.

**Comment by klimenttoshkov** [comment]
According to [https://github.com/symfony/workflow/blob/5.x/CHANGELOG.md](url):
```
5.2.0
Dispatch an event when the subject enters in the workflow for the very first time
```

is the reason why 5. is fixed too. So we have only point 3 open, which I think is not stopping you from releasing new version.

**Comment by zerodahero** [comment]
Ok, that makes sense, there's been a lot of changes around workflow entered starting around 4.4, and I'll confess I haven't been very good keeping up to date with them, so I appreciate your research, findings, and patience on this!

For the workflow diagram (point 3), I'm not seeing anything in the Symfony docs distinguishing workflow dumps from state machine dumps? The UML is the only diagram I was with the transitions marked on the arrows, otherwise they're always the squares.

**Comment by klimenttoshkov** [comment]
> Ok, that makes sense, there's been a lot of changes around workflow entered starting around 4.4, and I'll confess I haven't been very good keeping up to date with them, so I appreciate your research, findings, and patience on this!

Sure, it is beneficial for me, this way I am studying the package in details.

> For the workflow diagram (point 3), I'm not seeing anything in the Symfony docs distinguishing workflow dumps from state machine dumps? The UML is the only diagram I was with the transitions marked on the arrows, otherwise they're always the squares.

Yes, about that diagram. Why it is different? I thought it may be related to "state machine". I may be wrong of course.

**Comment by zerodahero** [comment]
I think the diagram is just a different type, since it's UML vs GraphViz's simplified format. I can look into adding the UML dumper as a possibility if that's something that would be helpful?

**Comment by klimenttoshkov** [comment]
Two things would be useful:
1. Dump in current Workflow version does not require model anymore, so i guess you can update accordingly;
2. I would love to be able to use other renderers except dot, perhaps you could add options/switches for renderer selection (thad could include UML dumper as well)?

**Comment by zerodahero** [comment]
Sounds good! I've added those to the project board to try and take a look at, soon: https://github.com/zerodahero/laravel-workflow/projects/1

Closing this issue since it seems like the original issue has been addressed in release 3.2.1 Feel free to reopen if I missed anything!
